src/big/big_int.cr

```diff
@@ -410,36 +410,92 @@ struct BigInt < Int
   # TODO: check hash equality for numbers >= 2**63
   def_hash to_i64!
 
-  # Returns a string representation of self.
-  #
-  # ```
-  # require "big"
-  #
-  # BigInt.new("123456789101101987654321").to_s # => 123456789101101987654321
-  # ```
-  def to_s : String
-    String.new(to_cstr)
+  def to_s(base : Int = 10, *, precision : Int = 1, upcase : Bool = false) : String
+    raise ArgumentError.new("Invalid base #{base}") unless 2 <= base <= 36 || base == 62
+    raise ArgumentError.new("upcase must be false for base 62") if upcase && base == 62
+    raise ArgumentError.new("Precision must be non-negative") unless precision >= 0
+
+    case {self, precision}
+    when {0, 0}
+      ""
+    when {0, 1}
+      "0"
+    when {1, 1}
+      "1"
+    else
+      count = LibGMP.sizeinbase(self, base).to_i
+      negative = self < 0
+
+      if precision <= count
+        len = count + (negative ? 1 : 0)
+        String.new(len + 1) do |buffer| # null terminator required by GMP
+          LibGMP.get_str(buffer, upcase ? -base : base, self)
+          base62_swapcase(Slice.new(buffer, len)) if base == 62
+          {len, len}
+        end
+      else
+        len = precision + (negative ? 1 : 0)
+        String.new(len + 1) do |buffer|
+          # e.g. precision = 13, count = 8
+          # "_____12345678\0" for positive
+          # "_____-12345678\0" for negative
+          LibGMP.get_str(buffer + precision - count, upcase ? -base : base, self)
+
+          if negative
+            buffer.value = '-'.ord.to_u8
+            Intrinsics.memset(buffer + 1, '0'.ord.to_u8, precision - count, false)
+          else
+            Intrinsics.memset(buffer, '0'.ord.to_u8, precision - count, false)
+          end
```

Comment:
I think this reads better with a single call to `memset`:
```suggestion
          if negative
            buffer.value = '-'.ord.to_u8
            buffer += 1
          end
          
          Intrinsics.memset(buffer, '0'.ord.to_u8, precision - count, false)
```
The value of `buffer` gets changed but it only affects `base62_swapcase` which can simply be moved before the precision fill and limited to the slice filled by `get_str`.